Fix perfect mod incorrectly failing in some scenarios

smoogipoo opened the pull request:
Prereqs:
- [x] https://github.com/ppy/osu/pull/8066
- [x] new framework release

Fixes https://github.com/ppy/osu/issues/8078

smoogipoo committed:
Add failing test scenes

smoogipoo committed:
Fix ruleset nullref

smoogipoo committed:
Better check for failing state

smoogipoo committed:
Fix taiko infinity health drain on some beatmaps

smoogipoo committed:
Make default taiko HP 1 for test scene

smoogipoo committed:
Fix custom rulesets not being testable

smoogipoo committed:
Fix perfect mod failure cases

bdach commented on osu.Game.Rulesets.Catch.Tests/TestSceneCatchModPerfect.cs:
```diff
@@ -0,0 +1,54 @@
+
+        [TestCase(false)]
+        [TestCase(true)]
+        public void TestTestJuiceStream(bool shouldMiss)
```
```suggestion
        public void TestJuiceStream(bool shouldMiss)
```

bdach commented on osu.Game/Tests/Visual/ModPerfectTestScene.cs:
```diff
@@ -0,0 +1,68 @@
+        protected class HitObjectTestCase
+        {
+            public readonly HitObject HitObject;
+            public readonly bool FailOnMiss;
```
Do you have anything in mind for this for the future? As far as I can see this is always `true`, and I can't really see much usage opportunity in having this.

smoogipoo replied:
It's actually `false` in one case in `TestSceneCatchModPerfect`. Taiko should do the same, however it's a little hard to test drumroll ticks by themselves right now.

smoogipoo replied:
Same with spinner bonus ticks when they're implemented.

bdach replied:
Ah, my bad. Gotcha.

bdach commented on osu.Game.Rulesets.Osu.Tests/TestSceneOsuModPerfect.cs:
```diff
@@ -0,0 +1,52 @@
+using osu.Game.Tests.Visual;
+using osuTK;
+
+namespace osu.Game.Rulesets.Osu.Tests
```
For consistency with `TestSceneOsuModDifficultyAdjust`:
```suggestion
namespace osu.Game.Rulesets.Osu.Tests.Mods
```
Goes for all of the others if we're making this a convention.

bdach reviewed (commented):
A few nits from source. Haven't ran this with framework checkout yet but it compiles in my head.

smoogipoo committed:
Adjust namespaces

smoogipoo committed:
Fix test name

peppy committed:
Merge branch 'master' into fix-perfect-judgements

peppy committed:
Update to follow new naming/structure

bdach commented:
CI is failing, because at the end of tests, upon gameplay completion there is an attempt to import the score:

https://github.com/ppy/osu/blob/fd16d24362ac1ed98f05e9b6942475bc9962c639/osu.Game/Screens/Play/Player.cs#L618-L625

and sqlite is eating a foreign key exception (haven't checked exactly why, but I'd bet it's due to violating constraints on either `BeatmapSetInfo` or `BeatmapInfo`).

Seeing that `ReplayPlayer` is a thing already I'd say it might be best to allow `Player` inheritors to specify their own logic on completion, defaulting to import.

peppy commented:
I agree with that direction. If you're already deep enough that you can easily commit the fix as proposed, you're welcome to go ahead!

bdach commented:
So I started to implement that but then found these and facepalmed *very* hard at my blindness:

https://github.com/ppy/osu/blob/9c5e3942adee2afe2d49394ef7081c7a622ce32c/osu.Game/Screens/Play/Player.cs#L107-L111

Sort of eh that these flags are so easy to miss but I guess if they're there already might as well set them and call it a day?

peppy commented:
It's weird that we do flags in three ways (ctor / settable props / overridden virtual props). Makes discovery hard.

bdach committed:
Merge branch 'master' into fix-perfect-judgements

bdach committed:
Do not transition to result screen

bdach committed:
Fix player instantiation

Since ModTestScene.CreatePlayer would apply mods in addition to
instantiating the player, overriding it could lead to mistakenly also
overriding the code that was supposed to set up the test via
currentTestData.

Make ModTestScene.CreatePlayer sealed, which ensures that mod &
autoplay changes are applied, and expose ModTestScene.CreateModPlayer
instead which has the expected semantics.

bdach commented:
I passed in that flag and also fixed up another issue that would cause the tests to not do what they intended to (explanation in commit msg of c803de2). Seems good otherwise 👍

peppy commented on osu.Game/Tests/Visual/ModPerfectTestScene.cs:
```diff
@@ -0,0 +1,67 @@
+            }
+        }
+
+        protected class HitObjectTestCase
```
i still don't like this and think it's convoluted. should look at a better solution before merging (even if that means combining with the other "info" class in the parent test scene)

peppy replied:
just going to go with this for now (just renamed to match `ModTestScene`). i think we can refactor/rethink as this continues to get used in more places.

peppy committed:
Merge branch 'master' into fix-perfect-judgements

peppy committed:
Rename data class
